File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/segment/SegmentSoftmax.java
/*
 *  ******************************************************************************
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.transforms.custom.segment;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Segment-wise softmax: softmax over each contiguous run of elements sharing a segment ID.
 *
 * <p>Used in batched GAT attention to normalize attention scores within each graph in a
 * block-diagonal batch (where segmentIds = batchVec from {@link org.nd4j.linalg.api.ops.impl.sparse.GraphDisjointUnion}).
 *
 * <p>For 1D logits:
 * <pre>
 *   out[i] = exp(logits[i]) / sum_{j in seg(i)} exp(logits[j])
 * </pre>
 *
 * <p>C++ op name: {@code segment_softmax}
 * <ul>
 *   <li>Inputs: logits[N,...] (float), segmentIds[N] (INT32, sorted)</li>
 *   <li>IArgs: K (number of segments)</li>
 *   <li>Output: out[N,...] (same shape and dtype as logits)</li>
 * </ul>
 */
public class SegmentSoftmax extends DynamicCustomOp {

    private final long K;

    public SegmentSoftmax() { this.K = 0; }

    /**
     * SameDiff constructor.
     *
     * @param sd         the SameDiff graph
     * @param logits     [N] or [N,...] float attention logits
     * @param segmentIds [N] INT32 sorted segment IDs in [0, K)
     * @param K          number of segments
     */
    public SegmentSoftmax(SameDiff sd, SDVariable logits, SDVariable segmentIds, long K) {
        super(sd, new SDVariable[]{logits, segmentIds}, false);
        this.K = K;
        addIArgument(K);
    }

    /**
     * Eager (INDArray) constructor.
     */
    public SegmentSoftmax(INDArray logits, INDArray segmentIds, long K) {
        super(new INDArray[]{logits, segmentIds}, null);
        this.K = K;
        addIArgument(K);
    }

    @Override
    public String opName() { return "segment_softmax"; }

    @Override
    public int getNumOutputs() { return 1; }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        return Collections.singletonList(dataTypes.get(0));
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> grads) {
        SDVariable gradOut    = grads.get(0);
        SDVariable fwdOut     = outputVariables()[0];
        SDVariable logitsArg  = arg(0);
        SDVariable segIdsArg  = arg(1);

        SDVariable dLogits = new SegmentSoftmaxBp(sameDiff, logitsArg, segIdsArg, fwdOut, gradOut, K)
                .outputVariable();
        return Arrays.asList(dLogits, sameDiff.zerosLike(segIdsArg));
    }
}

File: libnd4j/include/ops/declarable/generic/parity_ops/segment_softmax.cpp
/* ******************************************************************************
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// segment_softmax — softmax within segments (for batched GAT attention).
//
// Inputs:
//   [0] logits     [N] or [N, d1, ...]  float  — attention logits
//   [1] segmentIds [N]                  INT32  — sorted segment IDs (0-based, in [0,K))
// IArgs:
//   [0] K — number of segments
// Output:
//   [0] out  — same shape as logits
//
// segment_softmax_bp — backward.
//
// Inputs:
//   [0] logits     [N, ...]  (shape only needed)
//   [1] segmentIds [N]       INT32
//   [2] out        [N, ...]  forward output
//   [3] gradOut    [N, ...]  upstream gradient
// IArgs:
//   [0] K
// Output:
//   [0] dLogits [N, ...]
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_segment_softmax)

#include <ops/declarable/headers/parity_ops.h>
#include <ops/declarable/helpers/segment_softmax.h>

namespace sd {
namespace ops {

// ────────────────────────────────────────────────────────────────────────────
// Forward
// ────────────────────────────────────────────────────────────────────────────

CUSTOM_OP_IMPL(segment_softmax, 2, 1, false, 0, 1) {
    auto logits     = INPUT_VARIABLE(0);
    auto segmentIds = INPUT_VARIABLE(1);
    auto out        = OUTPUT_NULLIFIED(0);
    const LongType K = INT_ARG(0);

    REQUIRE_TRUE(segmentIds->isVector(), 0,
                 "segment_softmax: segmentIds must be 1D, got rank %d", segmentIds->rankOf());
    REQUIRE_TRUE(segmentIds->lengthOf() == logits->sizeAt(0), 0,
                 "segment_softmax: segmentIds length (%lld) must equal logits.shape[0] (%lld)",
                 (long long)segmentIds->lengthOf(), (long long)logits->sizeAt(0));
    REQUIRE_TRUE(K >= 1, 0, "segment_softmax: K must be >= 1, got %lld", (long long)K);

    helpers::segmentSoftmax(K, *logits, *segmentIds, *out);
    return Status::OK;
}

DECLARE_SHAPE_FN(segment_softmax) {
    return SHAPELIST(CONSTANT(inputShape->at(0)));
}

DECLARE_TYPES(segment_softmax) {
    getOpDescriptor()
        ->setAllowedInputTypes(0, {ALL_FLOATS})
        ->setAllowedInputTypes(1, {ALL_INTS})
        ->setAllowedOutputTypes({ALL_FLOATS})
        ->setSameMode(false);
}

// ────────────────────────────────────────────────────────────────────────────
// Backward
// ────────────────────────────────────────────────────────────────────────────

CUSTOM_OP_IMPL(segment_softmax_bp, 4, 1, false, 0, 1) {
    auto logits     = INPUT_VARIABLE(0);
    auto segmentIds = INPUT_VARIABLE(1);
    auto fwdOut     = INPUT_VARIABLE(2);
    auto gradOut    = INPUT_VARIABLE(3);
    auto dLogits    = OUTPUT_NULLIFIED(0);
    const LongType K = INT_ARG(0);

    helpers::segmentSoftmaxBp(K, *logits, *segmentIds, *fwdOut, *gradOut, *dLogits);
    return Status::OK;
}

DECLARE_SHAPE_FN(segment_softmax_bp) {
    return SHAPELIST(CONSTANT(inputShape->at(0)));
}

DECLARE_TYPES(segment_softmax_bp) {
    getOpDescriptor()
        ->setAllowedInputTypes(0, {ALL_FLOATS})
        ->setAllowedInputTypes(1, {ALL_INTS})
        ->setAllowedInputTypes(2, {ALL_FLOATS})
        ->setAllowedInputTypes(3, {ALL_FLOATS})
        ->setAllowedOutputTypes({ALL_FLOATS})
        ->setSameMode(false);
}

}  // namespace ops
}  // namespace sd
#endif  // NOT_EXCLUDED(OP_segment_softmax)
